### PR TITLE
Ensure setting of invoice status in all invocing error scenarios

### PR DESCRIPTION
**Pull request by reggieriser**
## Description

This PR attempts to ensure that we set an invoice's status correctly in all invoicing error scenarios, not just upon a non-200 response code as we are currently doing.

## Reviewer Notes

Several sections of code were converted to services as part of this PR:
* `ProcessInvoice` -- handles generating the EDI string, sending it to GEX, and setting the invoice and line item state appropriately.
* `StoreInvoice` -- moves that code from the `ediinvoice` package to a service to prevent circular dependencies.

The story initially talked about using `defer` in Go, but that code (which you can see in the commit history as the first attempt) ended up being fairly hard to follow.  The code was refactored to a service in such a way that `defer` became unnecessary.

I did add one new Go library ([go-multierror](https://github.com/hashicorp/go-multierror)) to support the multiple errors that could be generated as a result of cascading failures in the `ProcessInvoice` service.

## Setup

There was a new Go dependency added, so make sure your dependencies are up-to-date with `make deps`.

It's a little tricky to test the error scenarios.  I've added some tests to try to simulate some of them, so do a `make server_test` to see if they run as expected.  

You can also edit the `ProcessInvoice` service directly in key places if you'd like to arbitrarily throw errors to see how the invoice status responds.

Of course, normal invoicing should still work as expected (you can use our usual `ENTDEL` shipment in the E2E data to run through that scenario).

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in [Querying the Database Safely](./docs/backend.md#querying-the-database-safely) have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/162960651) for this change

### CODE REVIEW

**Comment by stangah** [comment]
I find the `defer` style of cleanup kind of hard to read with all the logic going on in there, and it still leaves a bunch in the handler that should probably be a service. I threw together a quick commit showing how I was thinking about going about it: https://github.com/transcom/mymove/commit/c92e0a568aab8b0e8d6375dd0be551bd7d2ae8c0

It definitely doesn't work (or probably even compile), but the idea is to more literally wrap all the operations that could go wrong with something that updates the invoice if it goes wrong. Maybe there's some way to incorporate both?

**Comment by reggieriser** [comment]
@stangah, I agree that `defer` got fairly complicated in this case.  Take a look when you have a moment at my latest changes that are modeled after your example code.  `defer` is gone entirely -- not sure I see any reason to keep it the way things are structured now.

BTW, I ended up pulling in `go-multierror` to help with the multiple errors possible in some scenarios.  If we don't want the additional dependency, I can try to write something similar or just concatenate the error strings together as one.

I still need to work on tests and getting the latest master changes incorporated.

**Comment by stangah** [comment]
Thanks for refactoring! It looks great, that `multierror` package is neat

**Comment by reggieriser** [comment]
Merged in some code from master that led to a circular dependency given the previous refactoring of the handler code to a service.  So, I refactored the `StoreInvoice858C` code into a service too to get around the circular dependency caused by it being in the `ediinvoice` package.

**Comment by reggieriser** [comment]
The tests have been added now, so I've removed the WIP status.  This is now ready for official review.

File: pkg/service/invoice/process_invoice_test.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	suite.Equal(expectedStatus, invoice.Status)
+}
+
+func helperCheckLineItemInvoiceID(suite *InvoiceServiceSuite, shipmentID *uuid.UUID, invoiceID *uuid.UUID) {
+	lineItems, err := models.FetchLineItemsByShipmentID(suite.DB(), shipmentID)
+	suite.FatalNoError(err)
+
+	for _, lineItem := range lineItems {
+		if invoiceID == nil {
+			suite.Nil(lineItem.InvoiceID)
```
Comment (jacquelineIO) [inline pkg/service/invoice/process_invoice_test.go]: Why are some allowed to be NIL and others have an expected ID? It is because if they are accepted then they won't be included in the invoice?

Comment (reggieriser) [reply pkg/service/invoice/process_invoice_test.go]: Line items are only associated with an invoice (by populating a line item's `InvoiceID` field) if the invoice submission was successful.  Three of these tests are testing intentional error scenarios, so I'm making sure there is no associated invoice with any of the line items afterwards (i.e., `InvoiceID` is nil).  One of them is testing a successful invoice process so the associated line items should have a non-nil `InvoiceID`.

File: pkg/service/invoice/process_invoice.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// ProcessInvoice is a service object to generate/send/record an invoice.
+type ProcessInvoice struct {
+	DB                    *pop.Connection
+	GexSender             gex.SendToGex
+	SendProductionInvoice bool
+}
+
+// Call processes an invoice by generating the EDI, sending the invoice to GEX, and recording the status.
+func (p ProcessInvoice) Call(invoice *models.Invoice, shipment models.Shipment) (*string, *validate.Errors, error) {
```
[the remark this answers is not in the document]
Comment (reggieriser) [reply pkg/service/invoice/process_invoice.go]: This service may be a little unique in that it needs to handle those invoice processing pieces as a unit in order to update the invoice record's status appropriately if any part of that process fails.  It even calls some other services as part of its work, so you could consider it as sort of a higher-level abstraction of processing an invoice, and therefore does do a single logical unit of work.  I think where it gets more problematic is if I added another distinct piece of functionality to the same `ProcessInvoice` service that did some different unit of work.  @mikena-truss may have some thoughts on this.

That said, I'm sure we could do some further refactoring of services here, but I just wanted to capture the main flow for now as suggested earlier by @stangah .

File: pkg/service/invoice/process_invoice.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		verrs, err := p.updateInvoiceFailed(invoice, validate.NewErrors(), err)
+		return ediString, verrs, err
+	}
+
+	// Update invoice record as submitted
+	verrs, err := UpdateInvoiceSubmitted{DB: p.DB}.Call(invoice, shipment.ShipmentLineItems)
+	if err != nil || verrs.HasAny() {
+		// Updating as submitted failed, so we need to try to mark it as failed (which could fail too).
+		verrs, err := p.updateInvoiceFailed(invoice, verrs, err)
```
[the remark this answers is not in the document]
Comment (reggieriser) [reply pkg/service/invoice/process_invoice.go]: I just added another status for `InvoiceStatusUPDATEFAILURE` to attempt to capture the scenario where the invoice submission went through fine but then updating the status/line items failed.  So, we should now be able to distinguish between a failed invoice submission and a failed status update (assuming the saving of the failed status update actually succeeds).
